Timestamps for internal events

jomuel opened the pull request:
Add a timestamp column to the database to record the time when internal events are logged.

Solves #2287.

rakanalh commented:
Reviewing this PR

rakanalh commented on raiden/transfer/architecture.py:
```diff
@@ -164,12 +164,17 @@ class Event:
             recipient: Address,
             channel_identifier: ChannelID,
             message_identifier: MessageID,
+            log_time: int = None,
```
I am looking at this and thinking ahead and i have several comments:
1. None of the places in which we instantiate events will provide this, hence the None default argument. Which makes this redundant and unnecessary (code-wise).
2. We store the `log_time` in a separate field in the database, but we also store a `log_time` field as part of the serialized events which is also redundant (data-wise).
3. Given #2, once  the deserialization happens, what you could do is "inject" the `log_time` attribute into the deserialized events. Since we don't have access modifiers in Python, it should be fine if we do this. So i would revert all the changes that you did on all events.

One more note about this which is that when we move to using `dataclasses`, the base dataclass could define the log_time field so that it's set to `None` by default and some value when deserialization takes place.

jomuel replied:
The `to_dict` methods of the different events only add `log_time` to the dict representation if it has not been set, which should never happen. So the redundancy in the database (#2) will not occur.

I thought of 'injecting' a log_time attribute too but this approach looked safer and more readable to me. (If the API is given an event with no log-time for some reason, it should display it as such and not crash with "no such attribute"). Though another way to achieve this would be to make `log_time` a property of `Event` and read it with `getattr` internally.

rakanalh replied:
Here's how i think the implementation should have been done:
1. Introduce log_time field into the DB
2. Whenever an event is being stored, we set the timestamp field with the current time in ms.
3. When deserializing, we set the `log_time` field of those events.
4. an API call to fetch internal events will call [this function](https://github.com/raiden-network/raiden/blob/master/raiden/api/rest.py#L825) which will make sure each event returned will have the field injected.
The above means that the frontend will never receive a payload with the `log_time` field missing.

As for the redundancy in data, what i meant was the fact that we will have 2 places which may contain the same value... the `log_time` field in the DB row and the `log_time` field in the serialized event instance stored in `data` field.

jomuel replied:
Ok. That's what I did, except that I made `log_time` a regular attribute of the `Event` subclasses. (Which would not be serialized and put in the database as `to_dict` skips it if it is None). You are right that it is unnecessary and I will remove it.

rakanalh commented on raiden/transfer/architecture.py:
```diff
@@ -164,12 +164,17 @@ class Event:
 
 
-class SendMessageEvent(Event):
+class EventWithLogTime(Event):
```
I see the point of creating the marker... but it would have made more sense if we:
1. Define common attributes, i.e `log_time` and reuse this in child classes, OR
2. In [your WAL change](https://github.com/raiden-network/raiden/pull/2381/files#diff-34fcdfdd66b47942cc02bf2a12fa4925R173) we check if `isinstance(event, EventWithLogTime)` and then set the attribute value.
However, the marker in it's current form doesn't do anything which means introducing it is not necessary. `Event` could still serve as the base class if we're just going to set `log_time` for all events.

LefterisJP commented:
Also @jomuel please add some unit tests

jomuel commented on raiden/storage/wal.py:
```diff
@@ -1,10 +1,15 @@
 
 
+def _timestamp_ms():
+    return int(time.time() * 1000)
```
We don't have a specific use case in mind for this, so why not include the extra information and use milliseconds. We are just adding these timestamps in case some API client should want them in the future - If some raiden gui wants to display internal events being handled it will look cooler with exact times...

hackaugusto replied:
Why not https://en.wikipedia.org/wiki/ISO_8601 instead of a unix epoch?

jomuel commented:
I added one more unit test. Unfortunately not many interesting things can be tested here with unit tests (testing what happens in the api would require a `RaidenService`. I addressed the other requested changes but we still need a decision on the time format.

hackaugusto commented on docs/rest_api.rst:
```diff
@@ -804,6 +807,7 @@ Now for the internal Raiden events:
 
     [{
         'event': 'SendLockedTransfer',
+        'log_time': 153722455821,
```
please update this :)

hackaugusto commented on raiden/storage/sqlite.py:
```diff
@@ -176,7 +177,12 @@ def get_statechanges_by_identifier(self, from_identifier, to_identifier):
+            self,
+            limit: int = None,
+            offset: int = None,
+            wrap_with_timestamps: bool = False,
```
I would always wrap instead of using this flag

LefterisJP commented:
All tests passed except for the flaky stress test for which an issue is already open.

Merging ...
